triggers/service/auth/src/main/scala/com/daml/auth/middleware/oauth2/Cli.scala

```diff
@@ -0,0 +1,268 @@
+
+package com.daml.auth.middleware.oauth2
+
+import akka.http.scaladsl.model.Uri
+import com.auth0.jwt.algorithms.Algorithm
+import com.daml.auth.middleware.oauth2.Config.{
+  DefaultCookieSecure,
+  DefaultHttpPort,
+  DefaultLoginTimeout,
+  DefaultMaxLoginRequests,
+}
+import com.daml.cliopts
+import com.daml.jwt.{
+  ECDSAVerifier,
+  HMAC256Verifier,
+  JwksVerifier,
+  JwtVerifierBase,
+  JwtVerifierConfigurationCli,
+  RSA256Verifier,
+}
+import com.typesafe.scalalogging.StrictLogging
+import pureconfig.{ConfigReader, ConfigSource, ConvertHelpers}
+import pureconfig.error.ConfigReaderException
+import scopt.OptionParser
+
+import java.io.File
+import pureconfig.generic.semiauto._
+
+import java.nio.file.{Path, Paths}
+import scala.concurrent.duration
+import scala.concurrent.duration.FiniteDuration
+
+sealed trait ConfigError extends Product with Serializable {
+  def msg: String
+}
+case object MissingConfigError extends ConfigError {
+  val msg = "Missing auth middleware config file"
+}
+final case class ConfigParseError(msg: String) extends ConfigError
+
+case class Cli(
+    configFile: Option[File] = None,
+    // Host and port the middleware listens on
+    address: String = cliopts.Http.defaultAddress,
+    port: Int = DefaultHttpPort,
+    portFile: Option[Path] = None,
+    // The URI to which the OAuth2 server will redirect after a completed login flow.
+    // Must map to the `/cb` endpoint of the auth middleware.
+    callbackUri: Option[Uri] = None,
+    maxLoginRequests: Int = DefaultMaxLoginRequests,
+    loginTimeout: FiniteDuration = DefaultLoginTimeout,
+    cookieSecure: Boolean = DefaultCookieSecure,
+    // OAuth2 server endpoints
+    oauthAuth: Uri,
+    oauthToken: Uri,
+    // OAuth2 server request templates
+    oauthAuthTemplate: Option[Path],
+    oauthTokenTemplate: Option[Path],
+    oauthRefreshTemplate: Option[Path],
+    // OAuth2 client properties
+    clientId: String,
+    clientSecret: SecretString,
+    // Token verification
+    tokenVerifier: JwtVerifierBase,
+) {
+
+  import Cli._
+
+  def loadConfigFromFile: Either[ConfigError, Config] = {
+    require(configFile.nonEmpty, "Config file should be defined to load app config")
+    configFile
+      .map(f =>
+        try {
+          Right(ConfigSource.file(f).loadOrThrow[Config])
+        } catch {
+          case ex: ConfigReaderException[_] => Left(ConfigParseError(ex.failures.head.description))
+        }
+      )
+      .get
+  }
+
+  def loadConfigFromCliArgs: Config = {
+    val cfg = Config(
+      address,
+      port,
+      portFile,
+      callbackUri,
+      maxLoginRequests,
+      loginTimeout,
+      cookieSecure,
+      oauthAuth,
+      oauthToken,
+      oauthAuthTemplate,
+      oauthTokenTemplate,
+      oauthRefreshTemplate,
+      clientId,
+      clientSecret,
+      tokenVerifier,
+    )
+    cfg.validate
+    cfg
+  }
+}
+
+object Cli extends StrictLogging {
+  implicit val tokenVerifierReader =
```

Comment:
If you don't wish to annotate your implicits, make them `private[this]` insofar as that still compiles.

Reply:
added type annotations